components/core/src/clp/ffi/utils.cpp

```diff
@@ -0,0 +1,230 @@
+    switch (character) {
+        // NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        case 0x08:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('b');
+            break;
+        case 0x09:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('t');
+            break;
+        case 0x0A:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('n');
+            break;
+        case 0x0C:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('f');
+            break;
+        case 0x0D:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('r');
+            break;
+        case 0x22:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('\"');
+            break;
+        case 0x5C:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('\\');
+            break;
+        // NOLINTEND(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        default: {
+            constexpr uint8_t cControlCharacter{0x1F};
+            if (cControlCharacter >= character) {
+                // Allocate 6 + 1 size buffer to format control characters as "\u00bb", with the
+                // last byte used by `snprintf` to append '\0'
+                constexpr size_t cControlCharacterBufSize{7};
+                std::array<char, cControlCharacterBufSize> buf{};
+                std::ignore = snprintf(buf.data(), buf.size(), "\\u00%02x", character);
+                escaped_string.append(buf.cbegin(), buf.cend() - 1);
+            } else {
+                escaped_string.push_back(static_cast<char>(character));
+            }
+            break;
+        }
+    }
+}
+
+auto is_valid_code_point(uint32_t code_point, size_t encoding_length) -> bool {
+    switch (encoding_length) {
+        // NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        case 1:
+            return code_point <= 0x7F;
+        case 2:
+            return (0x80 <= code_point && code_point <= 0x7FF);
+        case 3:
+            return (0x800 <= code_point && code_point <= 0xFFFF);
+        case 4:
+            return (0x1'0000 <= code_point && code_point <= 0x10'FFFF);
+        // NOLINTEND(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        default:
+            return false;
+    }
+}
+
+auto update_code_point(uint32_t code_point, uint8_t continuation_byte) -> uint32_t {
+    constexpr uint32_t cContinuationBytePayloadMask{0x3F};
+    constexpr uint8_t cNumContinuationBytePayloadBits{6};
+    return (code_point << cNumContinuationBytePayloadBits)
+           + (continuation_byte & cContinuationBytePayloadMask);
+}
+}  // namespace
+
+auto validate_and_escape_utf8_string(string_view raw) -> std::optional<string> {
```

Comment:
Before a detailed review:

Can we defer appending until we see an escaped character or we reach the end of the string? E.g., if there are no characters requiring escapes, we only need 1 append; if there are two chars, we need the escape + 2 appends; etc.

For 0-2 byte UTF-8 characters, we only need to validate the header bytes, right? Validating the code point is unnecessary since all possible values are valid. For a 3-byte UTF-8 character, we need to validate the header + the max range of the top 5 bits of the codepoint. If that's all correct, I think there's probably a more efficient way to write this.

Reply:
Discussed offline that not all possible values are valid since there's each code point's lower-bound is not 0.